### PR TITLE
SecureComms: Add support daemonConfig

### PR DESCRIPTION
**Pull request by davidhadas**
See:
 - https://github.com/confidential-containers/cloud-api-adaptor/issues/2096

Support configuring the APF Secure Comms from the CAA side including:
 - WN public Key
 - PP private key
 - Activating Secure Comms
 - inbouns and outbounds of th PP

This is useful for activating Secure Comms from the CAA and without Trustee. It can be used for Testing without producing dedicated podvms which activate Secure Comms and set Inbounds/Outbounds by default. It can also be used for non-Coco peerpods.

### CODE REVIEW

**Comment by davidhadas** [comment]
/hold
This is WIP.

**Comment by davidhadas** [comment]
cc: @bpradipt

File: src/cloud-api-adaptor/cmd/cloud-api-adaptor/main.go
Comment (stevenhorsman) [inline src/cloud-api-adaptor/cmd/cloud-api-adaptor/main.go:123]: Does this PR depend on https://github.com/confidential-containers/cloud-api-adaptor/pull/2073 where these changes also are?

File: src/cloud-api-adaptor/docs/SecureComms.md
Comment (stevenhorsman) [inline src/cloud-api-adaptor/docs/SecureComms.md:136]: I think we have the second issue here with repeated docs?

Comment (davidhadas) [reply src/cloud-api-adaptor/docs/SecureComms.md]: Correct - thanks, fixed.

**Comment by davidhadas** [comment]
@yoheiueda, @stevenhorsman, @bpradipt 

This PR adds the ability to configure SecureComms by auto-generating secrets, the same way it is done today by TLS. 
This is a necessary step to merge #2089 which is our first e2e test for SecureComms.

Since it is aligned with the existing TLS mechanisms and follow the same design and footsteps for the delivery of the secrets, I hope we can merge it without delay.  

@bpradipt has indicated that this PR aligns SecureComms with TLS such that all TLS use cases apparently are now also covered by SecureComms. This may lead to a further discussion about moving away from the less secure TLS option for maintainability (maybe best to do after we merge #2089).

File: src/cloud-api-adaptor/cmd/agent-protocol-forwarder/main.go
Comment (bpradipt) [inline src/cloud-api-adaptor/cmd/agent-protocol-forwarder/main.go:123]: nit: this statement can be removed

File: src/cloud-api-adaptor/cmd/agent-protocol-forwarder/main.go
Comment (bpradipt) [inline src/cloud-api-adaptor/cmd/agent-protocol-forwarder/main.go:118]: nit: CoCo in production -> CoCo

File: src/cloud-api-adaptor/cmd/agent-protocol-forwarder/main.go
Comment (bpradipt) [reply src/cloud-api-adaptor/cmd/agent-protocol-forwarder/main.go:123]: may be replaced with "// non-CoCo"

File: src/cloud-api-adaptor/cmd/cloud-api-adaptor/main.go
Comment (bpradipt) [reply src/cloud-api-adaptor/cmd/cloud-api-adaptor/main.go:146]: Strike that. So when secureComms is `true` and trustee is not there (aka Attestation Phase is optional), how is that handled?

File: src/cloud-api-adaptor/pkg/adaptor/cloud/cloud.go
Comment (bpradipt) [inline src/cloud-api-adaptor/pkg/adaptor/cloud/cloud.go:53]: Should this be CaaConfig ? `ServerConfig` is overloaded.

File: src/cloud-api-adaptor/pkg/adaptor/cloud/cloud.go
Comment (bpradipt) [inline src/cloud-api-adaptor/pkg/adaptor/cloud/cloud.go:380]: nit: ci -> sshCi (makes it clearer)

File: src/cloud-api-adaptor/pkg/securecomms/ppssh/ppsecrets.go
Comment (bpradipt) [inline src/cloud-api-adaptor/pkg/securecomms/ppssh/ppsecrets.go:43]: How are the WN and PP keys  derived when Trustee is not there ?

File: src/cloud-api-adaptor/cmd/cloud-api-adaptor/main.go
Comment (davidhadas) [reply src/cloud-api-adaptor/cmd/cloud-api-adaptor/main.go:146]: secureComms bool flag control if secureComms is enabled or not.

Currently the PR does not have a dedicated  bool flag to indicate "No attestation" mode. 
Instead, the PR reuse the `secure-comms-kbs` string flag. If that flag is set to "false", the inner `cfg.serverConfig.SecureCommsKbsAddress` is set to an empty string which is used internally to indicate that "No attestation" mode.

We can instead add a bool flag such as `secureCommsSkipAttestation`.

File: src/cloud-api-adaptor/pkg/adaptor/cloud/cloud.go
Comment (davidhadas) [reply src/cloud-api-adaptor/pkg/adaptor/cloud/cloud.go:53]: The `ServerConfig` type pre-dates this PR. 
It was moved from `src/cloud-api-adaptor/pkg/adaptor/server.go`. 

If we want to change it, we should do that in a separate PR.

File: src/cloud-api-adaptor/pkg/securecomms/ppssh/ppsecrets.go
Comment (davidhadas) [reply src/cloud-api-adaptor/pkg/securecomms/ppssh/ppsecrets.go:43]: when Trustee is not there, WN and PP keys are sent using deamonConfig and PpSecrets.SetKey() is used to set them into the PpSecrets object.

PpSecrets.Go() is never called in this case - so this code is only used when attestation is used.

Comment (bpradipt) [reply src/cloud-api-adaptor/pkg/securecomms/ppssh/ppsecrets.go:43]: Where is the WN and PP key generated before sending via daemonconfig ?

File: src/cloud-api-adaptor/cmd/cloud-api-adaptor/main.go
Comment (bpradipt) [reply src/cloud-api-adaptor/cmd/cloud-api-adaptor/main.go:146]: How about using empty string for `secure-comms-kbs` to infer no attestation is needed? A valid value will indicate attestation is required.

File: src/cloud-api-adaptor/pkg/securecomms/ppssh/ppsecrets.go
Comment (davidhadas) [reply src/cloud-api-adaptor/pkg/securecomms/ppssh/ppsecrets.go:43]: See func (s *cloudService) CreateVM() in cloud.go. 
This is where `daemonConfig.PpPrivateKey` and `daemonConfig.WnPublicKey` are set. 
PpPrivateKey comes from s.sshClient.InitPP() which is executed once per PP and where the private PP key is generated. 
WnPublicKey comes from s.sshClient.GetWnPublicKey() - originally generated once per cluster in InitSshClient()

File: src/cloud-api-adaptor/cmd/cloud-api-adaptor/main.go
Comment (davidhadas) [reply src/cloud-api-adaptor/cmd/cloud-api-adaptor/main.go:146]: I would prefer not to do that since this will not allow us to use the default "kbs-service.trustee-operator-system:8080" which are are using today. 
Today, if we `secure-comms-kbs` is not set, it is set to this default value which makes sense to anyone installing the system against a standard trustee-operator deployment.

Comment (bpradipt) [reply src/cloud-api-adaptor/cmd/cloud-api-adaptor/main.go:146]: I see.. In that case using a separate bool flag makes more sense and is intuitive. 
@stevenhorsman @yoheiueda wdyt? Should skipping attestation phase be a separate bool flag or setting the secure-comms-kbs addr to `false` is fine ?

File: src/cloud-api-adaptor/pkg/adaptor/cloud/cloud.go
Comment (bpradipt) [reply src/cloud-api-adaptor/pkg/adaptor/cloud/cloud.go:53]: Yeah it makes sense.

File: src/cloud-api-adaptor/docs/SecureComms.md
Comment (bpradipt) [inline src/cloud-api-adaptor/docs/SecureComms.md:57]: Some suggestions for the doc
1. Add a brief mention that this is an alternative to TLS encryption between the Adaptor and the Forwarder.
2. Add a mention about how the Attestation phase becomes optional if KBS is not used. Let's use the Attestation Phase (mandatory or optional) as a way to distinguish CoCo and Non-CoCo
3. Show two sequence of steps for SecureComms. One with Attestation Phase (and KBS). Another without the Attestation Phase. Use the same for setup instructions
4. If you can add a diagram showing a e2e example for inbound/outbound and namespaces from both the CAA and pod VM side, it'll be really helpful to visualise.

File: src/cloud-api-adaptor/cmd/cloud-api-adaptor/main.go
Comment (yoheiueda) [reply src/cloud-api-adaptor/cmd/cloud-api-adaptor/main.go:146]: A separate bool flag sounds more intuitive to me.

File: src/cloud-api-adaptor/docs/SecureComms.md
Comment (bpradipt) [inline src/cloud-api-adaptor/docs/SecureComms.md:32]:
```suggestion
As shown in the following diagram, using `SecureComms`, each Peer Pod is consequently served using a single communication channel implemented as an SSH channel. This single TCP session is initiated from the Worker Node (WN) and terminated at the Peer Pod (PP). It is used for any control communication between the WN and PP and vice versa. This consolidation of communication between a WN and PP via a single SSH channel improves the operational aspects of WN and PP communication, especially when WN and PP are in different networks. The WN network does not need to open ingress ports to expose services to PP. For example, when the communication between CDH (in PP) and the Trustee (in the WN network) is via the SSH channel, the Trustee service port does not need to be open to enable communication with CDH (PP).
```

File: src/cloud-api-adaptor/docs/SecureComms.md
Comment (bpradipt) [inline src/cloud-api-adaptor/docs/SecureComms.md:30]:
```suggestion
Once the SSH channel is created, clients at the Peer Pod side such as `Attestation Agent` connects to the services offered by the cluster side such as `KBS` using the channel. Likewise, clients located at the cluster side such as the `Kata shim`, connects to the services on the Peer Pod side such as the `Kata Agent` using the same channel.
```

File: src/cloud-api-adaptor/docs/SecureComms.md
Comment (bpradipt) [inline src/cloud-api-adaptor/docs/SecureComms.md:21]:
```suggestion
`Secure Comms` uses a Go SSH Lib to secure the communication between the cluster Worker Nodes (WNs) and Peer Pods (PPs). To communicate with the PPs, an SSH channel is created between the `cloud-api-adaptor` in the WN and the `agent-protocol-forwarder` in the PP. 
The WN is the client part of the channel, and the PP is the server part of the SSH channel. All communications between the PP and any services available on the WN side can be configured to use the SSH channel.
```

File: src/cloud-api-adaptor/docs/SecureComms.md
Comment (bpradipt) [inline src/cloud-api-adaptor/docs/SecureComms.md:7]:
```suggestion
Both `Secure Comms` and `TLS encryption` supports a method of delivering keys to encrypt the control-plane traffic between the `Adaptor` and the `Forwarder` via user-data during pod VM creation. Delivering the keys via the user-data may allow a malicious entity on the cloud provider side to gain access to the keys and modify any communication between the WNs and PPs. For this reason, `Secure Comms` also supports an alternate mode where the keys are delivered to the PP via `Trustee`. The keys are delivered to the PP after an attestation phase (See diagram below). Delivering the keys via `Trustee` protects the PP from a malicious entity on the cloud provider side. 
We name this mode of operation as `Trustee Mode`. The option of delivering keys via user-data is named  as `NoTrustee Mode`.
```

File: src/cloud-api-adaptor/docs/SecureComms.md
Comment (bpradipt) [inline src/cloud-api-adaptor/docs/SecureComms.md:47]:
```suggestion
Worker Node creates the Peer Pod VM and deliver the keys via daemonConfig as part of the userdata of the Cloud API.
```

File: src/cloud-api-adaptor/docs/SecureComms.md
Comment (bpradipt) [inline src/cloud-api-adaptor/docs/SecureComms.md:43]:
```suggestion
Worker Node establishes a "Kubernetes Phase" SSH channel with the Peer Pod VM - Both sides verify the identity of the other side using the keys delivered via Trustee.
```

File: src/cloud-api-adaptor/docs/SecureComms.md
Comment (bpradipt) [inline src/cloud-api-adaptor/docs/SecureComms.md:49]:
```suggestion
- Worker Node establishes a "Kubernetes Phase" SSH channel with the Peer Pod VM - Both sides verify the identity of the other side using the keys delivered via daemonConfig.
```

File: src/cloud-api-adaptor/docs/SecureComms.md
Comment (bpradipt) [inline src/cloud-api-adaptor/docs/SecureComms.md:51]: nit-> Runtime shim -> Kata shim

File: src/cloud-api-adaptor/cmd/cloud-api-adaptor/main.go
Comment (davidhadas) [reply src/cloud-api-adaptor/cmd/cloud-api-adaptor/main.go:146]: We can add this is a separate PR

Comment (bpradipt) [reply src/cloud-api-adaptor/cmd/cloud-api-adaptor/main.go:146]: @stevenhorsman @yoheiueda do you have any preference w.r.to adding the bool flag in this PR versus a new PR? Please see the thread above for context

Comment (davidhadas) [reply src/cloud-api-adaptor/cmd/cloud-api-adaptor/main.go]: I have now added a NoTrustee flag to replace setting the SecureCommsKbsAddress to false.

**Comment by stevenhorsman** [comment]
@davidhadas - can you rebase this PR please to pick up the check links fix. Thanks!

File: src/cloud-api-adaptor/docs/SecureComms.md
Comment (stevenhorsman) [inline src/cloud-api-adaptor/docs/SecureComms.md:7]: Can we replace WN for worker node and PP for peer pods please to help with clarity?

File: src/cloud-api-adaptor/docs/SecureComms.md
Comment (stevenhorsman) [inline src/cloud-api-adaptor/docs/SecureComms.md:49]: You've missed the bullet point here

File: src/cloud-api-adaptor/docs/SecureComms.md
Comment (stevenhorsman) [inline src/cloud-api-adaptor/docs/SecureComms.md:68]: In this set-up `SECURE_COMMS_NO_TRUSTEE: true` is required, so can we add into into the  first section and not the optional cm edit?

File: src/cloud-api-adaptor/docs/SecureComms.md
Comment (stevenhorsman) [reply src/cloud-api-adaptor/docs/SecureComms.md:7]: IMHO this would also improve the readability of the code too.

File: src/cloud-api-adaptor/docs/SecureComms.md
Comment (davidhadas) [reply src/cloud-api-adaptor/docs/SecureComms.md:68]: Done

**Review by stevenhorsman — APPROVED** [review]
This looks okay to me and all the tests on the PR pass, so I'm okay with it being merged.

**Review by bpradipt — APPROVED** [review]
/lgtm

**Pull request merged**
